torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -0,0 +1,192 @@
+from typing import Any, Dict, List, Tuple
+
+import torch
+import torch.utils._pytree as pytree
+from torch import Tensor
+from torch._C import DispatchKey
+from torch._ops import HigherOrderOperator
+from torch._prims_common import clone_preserve_strides
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+
+# NOTE: [auto-functionalizing custom ops]
+# Users may wish to torch.compile custom ops that mutate their inputs
+# and return nothing (this is a common class of operators).
+# torch.compile will automatically support this op without anyone needing
+# to provide a functionalization kernel for it. Here's how.
+#
+# Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> None
+# op. First, when FakeTensor sees this op, it can just return None.
+# This is the case because custom ops are not allowed to mutate input
+# metadata, so the FakeTensor rule for one that returns None is trivial.
+#
+# Next, when Python FunctionalTensor sees the op, it will functionalize
+# it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
+# HOP and replacing the mutated inputs with the outputs of this HOP.
+# This HOP effectively runs the functional version of the op when
+# called: it clones inputs that will be mutated, runs the op, and
+# then returns Tensors with the new values.
+
+
+class AutoFunctionalized(HigherOrderOperator):
+    """auto_functionalized(op, mutated_args_names, kwargs)
+
+    This HOP runs a "functional" version of op.
+
+    Concretely, it clones kwargs that `op` mutates (specified by
+    mutated_args_names), runs `op(**kwargs)` with the cloned values,
+    and then returns a flat Tuple of the cloned values that were mutated.
+
+    We have some restrictions on `op`, most notably that it returns None.
+    See `can_auto_functionalize` for the restrictions. We can likely lift
+    many of these if users request it.
+    """
+
+    def __init__(self):
+        super().__init__("auto_functionalized")
+
+    def __call__(
+        self,
+        op: torch._ops.OpOverload,
+        mutated_args_names: List[str],
+        kwargs: Dict[str, Any],
+    ) -> Tuple[Tensor, ...]:
+        assert can_auto_functionalize(op)
+        assert isinstance(mutated_args_names, list)
+        assert isinstance(kwargs, dict)
+        return super().__call__(op, mutated_args_names, kwargs)
+
+
+auto_functionalized = AutoFunctionalized()
+
+
+def can_auto_functionalize(op: torch._ops.OperatorBase) -> bool:
```

Comment:
Do you think it's returning False for ops in the ATen namespace? Just so we don't accidentally auto-functionalize them (even though I think in way you use this in `fake_tensor.py` today, that will never happen since you only try to run this during the fallback).

My thought was that this will be silently wrong for metadata-mutation ops - and even once you add op_tests to error on metadata-mutation, it's still valid to have aten ops that mutate metadata.